evaluation/LM_baseline.py

```diff
@@ -0,0 +1,151 @@
+import torch
+import math
+import time
+
+def predict_next_token(logits, top_k, tokenizer):
+    # Get top-k predicted token IDs and their probabilities
+    top_k_ids = torch.topk(logits, k=top_k, dim=-1)
+    top_k_token_probs = torch.softmax(top_k_ids.values, dim=-1)
+    
+    # Decode token IDs to tokens
+    tokens = [tokenizer.decode(idx.item()) for idx in top_k_ids.indices[0]]
+    
+    # Pair tokens with their probabilities
+    token_probabilities = list(zip(tokens, top_k_token_probs.tolist()[0]))
+    return token_probabilities
+
+def calculate_perplexity_ex(model, tokenizer, input_text, top_k=5):
+    inputs = tokenizer(input_text, return_tensors="pt")
+
+    input_ids = inputs["input_ids"].squeeze(0)
+    # print(input_ids)
+
+    total_loss = 0
+    total_tokens = 0
+
+    # Loop through all position
+    for i in range(len(input_ids)):
+        # Generate next token predictions
+        with torch.no_grad():
+            ids = input_ids[:i+1].unsqueeze(0)
+            sent = tokenizer.decode(ids[0])
+            # print(ids)
+            
+            outputs = model(input_ids = ids, labels = ids)
+            # print(outputs)
+        
+        # predictions = outputs.logits[:, -1, :]
+        loss = outputs.loss
```

Comment:
I think the loss might already be normalized by sentence length, so the normalization for perplexity calculation is repeated. You may need to double check.